src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,986 @@
+            {
+                // before foreach(var x in from a in b select a)
+
+                if (forEachStatement.Expression.WalkDownParentheses() != _source)
+                {
+                    documentUpdateInfo = default;
+                    return false;
+                }
+
+                // check that the body of the forEach does not contain any identifiers from the query
+                foreach (var identifierName in queryExpressionProcessingInfo.IdentifierNames)
+                {
+                    // Identifier from the foreach can already be in scope of the foreach statement.
+                    if (forEachStatement.Identifier.ValueText != identifierName)
+                    {
+                        if (_semanticFacts.GenerateUniqueName(
+                                _semanticModel,
+                                location: forEachStatement.Statement,
+                                containerOpt: forEachStatement.Statement,
+                                baseName: identifierName,
+                                usedNames: Enumerable.Empty<string>(),
+                                _cancellationToken).ValueText != identifierName)
+                        {
+                            documentUpdateInfo = default;
+                            return false;
+                        }
+                    }
+                }
+
+                // If query does not contains identifier with the same name as declared in the foreach,
+                // declare this identifier in the body.
+                if (!queryExpressionProcessingInfo.ContainsIdentifier(forEachStatement.Identifier))
+                {
+                    documentUpdateInfo = ConvertIfInToForeachWithExtraVariableDeclaration(forEachStatement, queryExpressionProcessingInfo);
+                    return true;
+                }
+                else
+                {
+                    // The last select expression in the query returns this identifier:
+                    // foreach(var thisIdentifier in from ....... select thisIdentifier)
+                    // if thisIdentifier in foreach is var, it is OK
+                    // if a type is specified for thisIdentifier in forEach, check that the type is the same as in the select expression
+                    // foreach(MyType thisIdentifier in from ....... from MyType thisIdentifier ... select thisIdentifier)
+
+                    // The last clause in query stack must be SelectClauseSyntax.
+                    var lastSelectExpression = ((SelectClauseSyntax)queryExpressionProcessingInfo.Stack.Peek()).Expression;
+                    if (lastSelectExpression is IdentifierNameSyntax identifierName &&
+                        forEachStatement.Identifier.ValueText == identifierName.Identifier.ValueText &&
+                        _semanticModel.GetSymbolInfo(identifierName, _cancellationToken).Symbol.Kind == SymbolKind.RangeVariable)
```

Comment:
>_semanticModel.GetSymbolInfo(identifierName, _cancellationToken).Symbol.Kind == SymbolKind.RangeVariable [](start = 24, length = 104)

Would it be better to simply look it up in queryExpressionProcessingInfo.IdentifierNames? I think the current check will let range variables from an enclosing query through, but we are probably interested only in range variables from the query we are refactoring.